### PR TITLE
More routes

### PR DESCRIPTION
**Pull request by zaucker**
- Add/fix some missing/broken routes
- Adapt frontend
- Adapt tests

### CODE REVIEW

File: lib/Agrammon/DB/Datasets.pm6 [hunk 1]
```diff
@@ -77,6 +77,22 @@ class Agrammon::DB::Datasets does Agrammon::DB {
...
 
+    method send(@datasets, $model, $new-username) {
+        my @cloned;
+        my $timestamp = ~DateTime.now( formatter => sub ($_) {
+            sprintf '%02d.%02d.%04d %02d:%02d:%02d',
+                    .day, .month, .year, .hour, .minute, .second,
+        });
+
+        for @datasets -> $old-dataset {
+            my $new-dataset = "$old-dataset - Kopie von " ~ $!user.username ~  "- $timestamp";
```
[the remark this answers is not in the document]
Comment (jnthn) [reply lib/Agrammon/DB/Datasets.pm6:88]: And maybe nicer with a space before the `-`?

File: lib/Agrammon/DB/Datasets.pm6 [hunk 2]
```diff
@@ -77,6 +77,22 @@ class Agrammon::DB::Datasets does Agrammon::DB {
...
+            Agrammon::DB::Dataset.new(:$!user, :$model).clone(:$new-username, :$old-dataset, :$new-dataset);
+            @cloned.push($old-dataset);
+        }
+        # expected by GUI
+        return %( :sent(@cloned.elems) );
```
Comment (jnthn) [inline lib/Agrammon/DB/Datasets.pm6:93]: If we only want the elems, can just `@dataset.elems`?

File: lib/Agrammon/OutputFormatter/Excel.pm6 [hunk 3]
```diff
@@ -97,22 +103,22 @@ sub input-output-as-excel(
...
+#        }
+#        if $instance and $instance ne $last-instance {
+#            $input-sheet-formatted.set($row-formatted, $col+1, $instance, :bold);
+#            $row-formatted++;
+#            $last-instance = $instance;
+#        }
+#        $input-sheet-formatted.set($row-formatted, $col+1, %rec<input>);
+#        $input-sheet-formatted.set($row-formatted, $col+2, (%rec<value-translated> // '???'), :number-format('#,#'), :horizontal-align(RightAlign));
+#        $input-sheet-formatted.set($row-formatted, $col+3, %rec<unit>);
+#        $row-formatted++;
```
Comment (jnthn) [inline lib/Agrammon/OutputFormatter/Excel.pm6:121]: Is this really meant to be in the PR?

File: lib/Agrammon/DB/Dataset.pm6 [hunk 4]
```diff
@@ -487,28 +497,30 @@ class Agrammon::DB::Dataset does Agrammon::DB {
...
-            #       AND data_var      LIKE $3
-            #       AND data_instance LIKE $4
-            #     SQL
-            #     $i++;
-            # }
+        my $rows-ordered = 0;
+        self.with-db: -> $db {
+            my $i = 0;
+            for (@instances) {
+                my $pattern  = $_;
```
[the remark this answers is not in the document]
Comment (zaucker) [reply lib/Agrammon/DB/Dataset.pm6:506]: Done.

File: lib/Agrammon/DB/Dataset.pm6 [hunk 5]
```diff
@@ -487,28 +497,30 @@ class Agrammon::DB::Dataset does Agrammon::DB {
...
+                my $instance = $1;
+                $var         = "$var\[\]%";
+                $db.query(q:to/SQL/, $i++, $username, $!name, $var, $instance);
+                UPDATE data_new SET data_instance_order = $1
+                 WHERE data_dataset = dataset_name2id($2,$3)
+                   AND data_var     LIKE $4
+                   AND data_instance = $5
+                RETURNING data_instance_order
+                SQL
+                $rows-ordered++;
```
[the remark this answers is not in the document]
Comment (zaucker) [reply lib/Agrammon/DB/Dataset.pm6:518]: Except if an `UPDATE` failed, was what I thought. But in this case there would be an exception. Changed the code to throw the right exception then.

File: lib/Agrammon/DB/Dataset.pm6 [hunk 6]
```diff
@@ -487,28 +497,30 @@ class Agrammon::DB::Dataset does Agrammon::DB {
...
+            }
         }
-        return True;
+
+        # reordering failed
+        die X::Agrammon::DB::Dataset::InstanceReorderFailed.new(:$!name) unless $rows-ordered == @instances.elems;
```
[the remark this answers is not in the document]
Comment (zaucker) [reply lib/Agrammon/DB/Dataset.pm6:523]: See above.

File: lib/Agrammon/DB/Datasets.pm6 (hunk 1 above)
[the remark this answers is not in the document]
Comment (zaucker) [reply lib/Agrammon/DB/Datasets.pm6:88]: > And maybe nicer with a space before the `-`?

Added.

[the remark this answers is not in the document]
Comment (zaucker) [reply lib/Agrammon/DB/Datasets.pm6:88]: > Needs to be language sensitive at some point?

Yes, but not before I switch to `.po` files. This was in German only for the last 10 years :-)

File: lib/Agrammon/DB/Datasets.pm6 [hunk 7]
```diff
@@ -77,6 +77,22 @@ class Agrammon::DB::Datasets does Agrammon::DB {
         return $deleted;
     }
 
+    method send(@datasets, $model, $new-username) {
+        my @cloned;
+        my $timestamp = ~DateTime.now( formatter => sub ($_) {
+            sprintf '%02d.%02d.%04d %02d:%02d:%02d',
+                    .day, .month, .year, .hour, .minute, .second,
+        });
```
[the remark this answers is not in the document]
Comment (zaucker) [reply lib/Agrammon/DB/Datasets.pm6:85]: Put it into a module `Agrammon::Timestamp`. Perhaps I should create a helper module.

File: lib/Agrammon/OutputFormatter/Excel.pm6 (hunk 3 above)
Comment (zaucker) [reply lib/Agrammon/OutputFormatter/Excel.pm6:121]: At the moment yes, will go back in once the testing is over (not needed there and later the performance impact doesn't matter so much). Added a comment about it.

**Pull request merged**